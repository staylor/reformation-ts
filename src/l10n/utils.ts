const cache = {};

export const getMessages = (locale: string, scope?: string) => {
  let path = locale;
  if (scope) {
    path = `${scope}/${locale}`;
  }

  if (cache[path]) {
    return cache[path];
  }

  let messages;

  try {
    messages = require(`./${path}.json`);
  } catch (e) {
    if (locale === 'en') {
      throw e;
    }
    messages = getMessages('en', scope);
  }
  cache[path] = messages;
  return messages;
};
