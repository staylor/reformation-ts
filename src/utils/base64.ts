// Client notes:
// https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/Base64_encoding_and_decoding
// The "Unicode Problem"

function base64EncodeUnicode(str: string) {
  // Server uses Buffer API
  if (typeof document === 'undefined') {
    return Buffer.from(str).toString('base64');
  }
  // In most browsers, calling btoa() on a Unicode string will cause a Character Out Of Range exception.
  // First we use encodeURIComponent to get percent-encoded UTF-8,
  // then we convert the percent encodings into raw bytes which can be fed into btoa.
  return btoa(
    encodeURIComponent(str).replace(/%([0-9A-F]{2})/g, (match, p1) =>
      String.fromCharCode(parseInt(`0x${p1}`, 16))
    )
  );
}

function base64DecodeUnicode(str: string) {
  // Server uses Buffer API
  if (typeof document === 'undefined') {
    return Buffer.from(str, 'base64').toString();
  }
  // Going backwards: from bytestream, to percent-encoding, to original string.
  return decodeURIComponent(
    atob(str)
      .split('')
      .map(c => `%${`00${c.charCodeAt(0).toString(16)}`.slice(-2)}`)
      .join('')
  );
}

export { base64EncodeUnicode as base64Encode, base64DecodeUnicode as base64Decode };
