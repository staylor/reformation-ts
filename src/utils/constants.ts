export const TOKEN_KEY = 'reformation_auth_token';
export const TITLE_TEMPLATE = '%s - Reformation';
export const COOKIE_EXPIRES = 7;
export const COOKIE_PATH = '/';
