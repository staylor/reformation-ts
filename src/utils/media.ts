export const uploadUrl = (destination: string, fileName: string) => {
  return `https://storage.googleapis.com/wonderboymusic/${destination}/${fileName}`;
};
