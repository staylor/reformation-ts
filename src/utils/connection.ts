import { base64Decode, base64Encode } from 'utils/base64';

const PREFIX = 'arrayconnection:';

export function cursorToOffset(cursor: string) {
  return parseInt(base64Decode(cursor).substring(PREFIX.length), 10);
}

export function offsetToCursor(offset: number) {
  return base64Encode(PREFIX + offset);
}

interface Node {
  id: string;
}

interface Edge<T> {
  node: T;
  cursor: string;
}

interface Connection<T> {
  edges: Edge<T>[];
}

interface Collection<T> {
  [key: string]: Connection<T>;
}

interface FetchMoreResult<T> {
  fetchMoreResult: Collection<T>;
}

/**
 * All instances of infinite scroll follow the same convention
 * - Relay-based pagination using the Connection spec
 * - Always loading more items after the last loaded item
 *
 * To avoid component inheritance, `this` must be bound when calling this function
 */
export function bindLoadMore(prop: string) {
  return (e: Event): Promise<Collection<Node>> => {
    e.preventDefault();

    const { fetchMore, variables } = this.props.data;

    const data = this.props.data[prop];

    return fetchMore({
      variables: {
        ...variables,
        after: data.pageInfo.endCursor,
      },
      updateQuery: (
        previousResult: Collection<Node>,
        { fetchMoreResult }: FetchMoreResult<Node>
      ) => {
        const mergedResult = {
          ...fetchMoreResult,
          [prop]: {
            ...fetchMoreResult[prop],
            edges: previousResult[prop].edges.concat(fetchMoreResult[prop].edges),
          },
        };

        return mergedResult;
      },
    });
  };
}
