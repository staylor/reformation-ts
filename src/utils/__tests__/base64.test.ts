import { base64Encode, base64Decode } from 'utils/base64';

describe('base64', () => {
  test('serialization', () => {
    const value = 'lshjgjh34t0934t(*^&*^*^*^*&^KHJHHKJ)';

    const encoded = base64Encode(value);
    expect(encoded).toMatchSnapshot();
    const decoded = base64Decode(encoded);
    expect(decoded).toEqual(value);
  });

  test('unicode', () => {
    const value = 'kj🎷hkj🚀whe80980🙏98235😴#$%^&*()🚂';

    const encoded = base64Encode(value);
    expect(encoded).toMatchSnapshot();
    const decoded = base64Decode(encoded);
    expect(decoded).toEqual(value);
  });
});
