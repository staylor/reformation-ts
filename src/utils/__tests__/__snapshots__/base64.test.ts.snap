// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base64 serialization 1`] = `"bHNoamdqaDM0dDA5MzR0KCpeJipeKl4qXiomXktISkhIS0op"`;

exports[`base64 unicode 1`] = `"a2rwn463aGtq8J+agHdoZTgwOTgw8J+Zjzk4MjM18J+YtCMkJV4mKigp8J+agg=="`;
